### ISSUE TITLE
Initial Discussion

### ISSUE THREAD

**Issue by psychocoderHPC** [ComputationalRadiationPhysics/contributing#2]
This is the discussion to the first CRP C++ coding style

- the style based on
  - [PIConGPU coding style](https://github.com/ComputationalRadiationPhysics/picongpu/wiki/Coding-Guide-Lines)
  - code style discussion in ALPAKA https://github.com/ComputationalRadiationPhysics/alpaka/issues/102

The style is a mix of both sources.

Goals:
- create a coding style which can used for all C++ projects in [ComputationalRadiationPhysics](https://github.com/ComputationalRadiationPhysics)
- the code should looks equal independent of the used editor, IDE ( incl. web based editors)
- it must be write/ read able without IDE features such as 
  - block collapse function
  - code highlighting
- break the 80 column limit should avoid as best as possible by design (new line rules)
- rules should be consistent ( to add linter or code beautifier support )
- with heavy template usage the code should stay readable

**Comment by psychocoderHPC** [comment]
**Why is this used in the draft and not that?**

- doxygen commands `@` or `\`
  - with google I cant find any pro or cons for on of the versions
  - after checking the documentation I found that two command s `@{` and `@}` can only used with `@`
  - -> `@` is used that all doxygen are constistent
-  `space ` vs. `tab` for indention
  - -> `space` is used that the code looks equal in all IDE's
- different rules for definitions and expressions
  - -> expressions get own rules to avoid lines of codes explosion
  - -> equations looks better

**Comment by psychocoderHPC** [comment]
@ComputationalRadiationPhysics/xrt-maintainers  @ComputationalRadiationPhysics/dsp-maintainer 
please enter this discussion.

**Comment by ax3l** [comment]
the head of the [PIConGPU wiki](https://github.com/ComputationalRadiationPhysics/picongpu/wiki/Coding-Guide-Lines) has some more open points regarding conventions specifically for meta programming and traits that we might want to covered at some point.

- Add something template meta programming specific: public typedefs like `ValueType` and `type` (boost resultOf)
  - `ValueType` is a public typedef for internal storage types (e.g. a single element of a field)
  - type traits should either define a `::result` or `::type` (the literature is not very unified in that)
  - what is necessary for a `resultOf` (see also: [this current problem](https://github.com/ComputationalRadiationPhysics/picongpu/blob/8397fb022b40f11647e439ff33f3bab3bb8cf66c/src/picongpu/CMakeLists.txt#L262-L275))

**Comment by BenjaminW3** [comment]
Should all discussions be within this thread (confusing) or should we open one issue per topic?
I want to add to point 7 that all Preprocessor Macros have to be prefixed with a project identifier to avoid duplications across projects. E.g. in alpaka all macros start with `ALPAKA_`.

**Comment by psychocoderHPC** [comment]
IMO a pull request currently to master is the best.

**Comment by ax3l** [comment]
one issue per topic is a very good idea.

**Comment by psychocoderHPC** [comment]
Note to myself, file rule missed:
`an empty line is the last line in a file`

**Comment by Flamefire** [comment]
In https://github.com/ComputationalRadiationPhysics/contributing/blob/master/codingGuideLines/cpp.md#17-template-class--struct-and-type-definitions
What is the meaning of

    for C++11 `using` new line after assign `=` mandatory ( definition is indented )

? The example does not help to understand that and the usage of the inline code tags seems to be odd. 

Similar in https://github.com/ComputationalRadiationPhysics/contributing/blob/master/codingGuideLines/cpp.md#12-function-and-method-definitions with the discussion of the `oc-token`s (what is that?) The description talks about new lines around parentheses (?) but there are none in the example. Maybe it is meant, that one can put arguments on own lines, in which case the first newline must be after `(` and the last before `)`

**Comment by ax3l** [comment]
that rule is removed

**Comment by ax3l** [comment]
pls open **new issues** per topic!

**Comment by psychocoderHPC** [comment]
Please read this to see why we need a defined coding style:
http://www.smashingmagazine.com/2012/10/why-coding-style-matters/

**Comment by j-stephan** [comment]
> after checking the documentation I found that two command s @{ and @} can only be used with @

This is not true (or it may have changed in the past 5 1/2 years). I'm using `\{` in bactria and it works as expected.
